### PR TITLE
8328614: hsdis: dlsym can't find decode symbol

### PR DESCRIPTION
**Pull request by robehn**
Hi, please consider.

[8327045](https://bugs.openjdk.org/browse/JDK-8327045) hide these symbols.
Tested with gcc and clang, and llvm and binutils backend.

I didn't find any use of the "DLL_ENTRY", so I removed it.

Thanks, Robbin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328614](https://bugs.openjdk.org/browse/JDK-8328614): hsdis: dlsym can't find decode symbol (**Bug** - P3)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18400/head:pull/18400` \
`$ git checkout pull/18400`

Update a local copy of the PR: \
`$ git checkout pull/18400` \
`$ git pull https://git.openjdk.org/jdk.git pull/18400/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18400`

View PR using the GUI difftool: \
`$ git pr show -t 18400`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18400.diff">https://git.openjdk.org/jdk/pull/18400.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18400#issuecomment-2009980370)

### CODE REVIEW

File: src/hotspot/cpu/x86/.nativeInst_x86.cpp.swp
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/.nativeInst_x86.cpp.swp:1]: This file should not have been committed :)

Comment (robehn) [reply src/hotspot/cpu/x86/.nativeInst_x86.cpp.swp:1]: Ops, thanks!

**Comment by robehn** [comment]
@magicus is this how you want us to export these symbols?

**Comment by magicus** [comment]
> is this how you want us to export these symbols?

Close but no cigar. :-)

Use `JNIEXPORT` instead, that is properly defined for this purpose and works on all compilers. You will need to also add:

```
#include "jni.h"
```

If this is not picked up correctly, let me know and I'll help you get the include paths correctly in the build.

**Comment by magicus** [comment]
Also, apologies for forgetting to check hsdis when I changed the visibility. :-(

**Comment by robehn** [comment]
> > is this how you want us to export these symbols?
> 
> Close but no cigar. :-)
> 
> Use `JNIEXPORT` instead, that is properly defined for this purpose and works on all compilers. You will need to also add:
> 
> ```
> #include "jni.h"
> ```
> 
> If this is not picked up correctly, let me know and I'll help you get the include paths correctly in the build.

It's stand alone library, should we really make it dependent on the JDK?

**Comment by magicus** [comment]
Unless you start calling JDK functions, you will not make a program less stand-alone by including jni.h. In this case, you will only use a compile-time definition.

Ideally, we should have had more general EXPORT definitions separate from the rest of the JNI code, but someone started doing things that way, and well, here we are, 25 years later and now JNIEXPORT is everywhere in the JDK source base. :( I'd say that access to JNIEXPORT is about 50% of the reason jni.h is included...

**Comment by theRealAph** [comment]
> > > is this how you want us to export these symbols?
> > 
> > 
> > Close but no cigar. :-)
> > Use `JNIEXPORT` instead, that is properly defined for this purpose and works on all compilers. You will need to also add:
> > ```
> > #include "jni.h"
> > ```

> >   
> > If this is not picked up correctly, let me know and I'll help you get the include paths correctly in the build.
> 
> It's stand alone library, should we really make it dependent on the JDK?

No. And neither should we compile or link it with "-fvisibility=hidden". That is the root of
this problem.

**Comment by magicus** [comment]
> And neither should we compile or link it with "-fvisibility=hidden". That is the root of this problem.

If you suggest that we should not compile hsdis with hidden visibility, I disagree. I have been working hard on unifying build of native libraries across the entire product, to fix holes where we have not used a consistent way of compiling and/or linking. There is no reason to tread hsdis differently. If I restore using hidden visibility as an option that all native libraries, except hsdis, must opt in to, then we are just back to square one, and suddenly someone will forget about it. Instead, now we set -fvisibility=hidden in configure so nobody can forget about it.

That was the more general argument for alinging compilation and linking flags and behavior across the product. Regarding the symbol visibility more specifically, there is also the question of consistency across platforms. On Windows, the behavior corresponding to "-fvisibility=hidden" is always turned on. Functions can only be exported if they are explicitly annotated in the source code (or specified otherwise to the linker). So we are in any case forced to export functions on Windows. 

Let's have a look at this patch. Currently we have code like:

```
#ifdef _WIN32
__declspec(dllexport)
#endif
```

Robbin proposes to change this to 

```
#if defined(_WIN32)
__declspec(dllexport)
#elif defined(_GNU_SOURCE)
__attribute__ ((visibility ("default")))
#endif
```

My counter-proposal was to replace it with just `JNIEXPORT`. Surely you can't say that is a worse solution?

**Comment by magicus** [comment]
(In fact, I think we have a problem everywhere in the code base where someone is using `__declspec(dllexport)` directly)

**Comment by robehn** [comment]
> (In fact, I think we have a problem everywhere in the code base where someone is using `__declspec(dllexport)` directly)

src/java.base/share/native/libzip/zlib/zconf.h:#      define ZEXPORT   __declspec(dllexport)

ZEXPORT is only defined on win32, so we must build with default on linux.

**Comment by magicus** [comment]
> src/java.base/share/native/libzip/zlib/zconf.h:# define ZEXPORT __declspec(dllexport)

zlib is third party source that is copied into our repo. I did not mean that what I said for that applies to it. (Also, we should not really export *any* functions from zlib -- we encapsulate it in our own native shared library. I might have to look into if we are doing this wrong on Windows. But that is a separate question from this PR.)

**Comment by robehn** [comment]
> > src/java.base/share/native/libzip/zlib/zconf.h:# define ZEXPORT __declspec(dllexport)
> 
> zlib is third party source that is copied into our repo. I did not mean that what I said for that applies to it. (Also, we should not really export _any_ functions from zlib -- we encapsulate it in our own native shared library. I might have to look into if we are doing this wrong on Windows. But that is a separate question from this PR.)

Okay, so the difference is that me and @theRealAph consider hsdis a thrid party library, although the JDK folks maintain it.

**Comment by magicus** [comment]
It is original code in the JDK library. zlib is copied in from elsewhere. That is what I mean by third party code.

The only reason hsdis is specially treated is due to the fact that the original implementation was based on binutils, which could not -- for legal reason -- be distributed as a compiled library alongside the JDK. That forced us to erect an artificial barrier.

There are multiple JBS issues trying to rectify this, with different approaches, to get a non-binutils hsdis library build and shipped as any other native library in the JDK.

**Comment by robehn** [comment]
How would you add jni.h ?

**Comment by theRealAph** [comment]
> > And neither should we compile or link it with "-fvisibility=hidden". That is the root of this problem.
> 
> If you suggest that we should not compile hsdis with hidden visibility, I disagree.

Yes, that's what I would do. 

> I have been working hard on unifying build of native libraries across the entire product, to fix holes where we have not used a consistent way of compiling and/or linking. There is no reason to tread hsdis differently. If I restore using hidden visibility as an option that all native libraries, except hsdis, must opt in to, then we are just back to square one, and suddenly someone will forget about it. Instead, now we set -fvisibility=hidden in configure so nobody can forget about it.

OK, OK! So please can we get this fix in?

> Robbin proposes to change this to
> 
> ```
> #if defined(_WIN32)
> __declspec(dllexport)
> #elif defined(_GNU_SOURCE)
> __attribute__ ((visibility ("default")))
> #endif
> ```
> 
> My counter-proposal was to replace it with just `JNIEXPORT`. Surely you can't say that is a worse solution?

JNIEXPORT is better, I guess, but it does mean that hsdis is no longer standalone, and IMO it should have a pathological dependency on some JVM header file. That's the problem here.

But really, whatever works is good with me. The last thing I want to do is delay this fix any further.

Robbin has asked "How would you add jni.h ?" Is anyone going to answer?

**Comment by magicus** [comment]
Ideally, it should just be enough to add `#import "jni.h"`. This will definitely work for capstone and llvm. However, for binutils we disable our standard C include paths, so this will fail. (I'm not sure if this is really necessary, but that is how we currently do it)

To solve it you can either set `HSDIS_TOOLCHAIN_DEFAULT_CFLAGS` to `true` for binutils, or you can add the include path to jni.h directly. The former would be the best, but in case that does not work, try adding `java.base:include` to `EXTRA_HEADER_DIRS`.

**Comment by magicus** [comment]
I apologize for the late reply. I've been just working spotty hours due to spring break.

**Comment by robehn** [comment]
Thanks, updated. (Since windows build seems to need HSDIS_TOOLCHAIN_DEFAULT_CFLAGS, just adding java.base:include seemed a good easy way)

**Review by magicus — APPROVED** [review]
Looks good to me now. Thanks for indulging me in doing it this way. :-)

**Comment by TheShermanTanker** [comment]
> > Since windows build seems to need HSDIS_TOOLCHAIN_DEFAULT_CFLAGS
> 
> Riiiight. That is since we use a completely different compiler, gcc instead of cl. (Which is probably the worst hack in all of the JDK build system...)

I plan on adding support for using gcc to compile hsdis on Windows sometime soon (Like proper support for gcc as a toolchain on Windows, at least enough to support hsdis, so people can freely use any gcc instead of only being restricted to Cygwin), what might be an elegant solution to this issue, if you have any in mind?

**Comment by magicus** [comment]
> I plan on adding support for using gcc to compile hsdis on Windows sometime soon (Like proper support for gcc as a toolchain on Windows, at least enough to support hsdis, so people can freely use any gcc instead of only being restricted to Cygwin), what might be an elegant solution to this issue, if you have any in mind?

The proper solution is, I think, to encapsulate toolchain variables/definitions into a package, and update the build system to not assume anything about toolchains. In a proper language, I'd create like a `class Toolchain` with everything the system would need to know to use a toolchain, and then pass a `Toolchain toolchain` to all invocations that need to do anything with a toolchain. Now this is not possible with make, but I hope it can be simulated using "fake namespaces" and something like a `$(TOOLCHAIN)_COMPILER` kind of setup.

That would require some major surgery of the build system, but I think it will come out better in the end.

After that is done, fixing so that hsdis/binutils can be compiled with gcc on Windows will be a trivial side-effect.

**Comment by theRealAph** [comment]
> I apologize for the late reply. I've been just working spotty hours due to spring break.

I apologize for my bad temper. 😀

Thanks to everyone working on this. I still think that hsdis ought not to have any dependencies on HotSpot, but I'm not going to be fanatical about it.

**Review by luhenry — APPROVED** [review]

**Comment by robehn** [comment]
Thank you @luhenry !

**Comment by magicus** [comment]
FYI: The 2-reviewer rule only applies to Hotspot (and some other specific parts of the code base), and not generally for JDK fixes. So you had been fine to push with just one reviewer.

**Comment by robehn** [comment]
Yea, I know, but if hsdis is not an external library I considered it part of hotspot as it is the only use, hence I want two reviews. :)

**Review by Hamlin-Li — APPROVED** [review]
Looks good.
Although I'm not the right person to review this, thanks for explanation and discussion.

**Pull request closed without merging**